src/back/hlsl/writer.rs

```diff
@@ -2577,3 +2790,139 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
         Ok(())
     }
 }
+
+pub(super) struct MatrixType {
+    pub(super) columns: crate::VectorSize,
+    pub(super) rows: crate::VectorSize,
+    pub(super) width: crate::Bytes,
+}
+
+pub(super) fn get_inner_matrix_data(
+    module: &Module,
+    handle: Handle<crate::Type>,
+) -> Option<MatrixType> {
+    match module.types[handle].inner {
+        TypeInner::Matrix {
+            columns,
+            rows,
+            width,
+        } => Some(MatrixType {
+            columns,
+            rows,
+            width,
+        }),
+        TypeInner::Array { base, .. } => get_inner_matrix_data(module, base),
+        _ => None,
+    }
+}
+
+/// Returns the matrix data if the access chain starting at `base`:
+/// - starts with an expression with resolved type of [`TypeInner::Matrix`] if `direct = true`
+/// - contains one or more expressions with resolved type of [`TypeInner::Array`] of [`TypeInner::Matrix`]
+/// - ends at an expression with resolved type of [`TypeInner::Struct`]
+pub(super) fn get_inner_matrix_of_struct_array_member(
+    module: &Module,
+    base: Handle<crate::Expression>,
+    func_ctx: &back::FunctionCtx<'_>,
+    direct: bool,
+) -> Option<MatrixType> {
+    let mut mat_data = None;
+    let mut array_base = None;
+
+    let mut current_base = base;
+    loop {
+        let mut resolved = func_ctx.info[current_base].ty.inner_with(&module.types);
+        if let TypeInner::Pointer { base, .. } = *resolved {
+            resolved = &module.types[base].inner;
+        };
+
+        match *resolved {
+            TypeInner::Matrix {
+                columns,
+                rows,
+                width,
+            } => {
+                mat_data = Some(MatrixType {
+                    columns,
+                    rows,
+                    width,
+                })
+            }
+            TypeInner::Array { base, .. } => {
+                array_base = Some(base);
+            }
+            TypeInner::Struct { .. } => {
+                if let Some(array_base) = array_base {
+                    if direct {
+                        return mat_data;
+                    } else {
+                        return get_inner_matrix_data(module, array_base);
+                    }
```

Comment:
This part is a bit tricky, `mat_data` might be `None` because the resolved type of the first chain item might have been an array of matrices which is fine if direct is false. I guess the else block could also say `mat_data.or(get_inner_matrix_data...` (just a small optimization).

Reply:
This part is a bit tricky, `mat_data` might be `None` because the resolved type of the first chain item might have been an array of matrices which is fine if direct is false. I guess the else block could also say `mat_data.or(get_inner_matrix_data...` (just a small optimization).

Reply:
Okay - so the distinction matters when you have an access of a component that *contains* a matrix, but not the matrix itself. In this case, `mat_data` would be `None`, since we would never encounter an `Expression` whose type is `Matrix`, but calling `get_inner_matrix_data` would return `Some`, since it walks to the bottom of the type.